src/main/java/com/google/edith/servlets/UserInsights.java

```diff
@@ -0,0 +1,194 @@
+      } catch(EntityNotFoundException | NullPointerException e )  {
+        return 0;
+      }
+  };
+  
+  public UserInsights(String userId) {
+    this.userId = userId;
+  }
+
+  /**
+   * Finds the UserStats entity corresponding to {@code userId} in datastore.
+   * @return UserStats entity for this user
+   */
+  private Entity retreiveUserStats() {
+    Filter idFilter = new FilterPredicate("UserId", 
+                                FilterOperator.EQUAL,
+                                userId);
+    Query query = new Query("UserStats").setFilter(idFilter);
+    PreparedQuery results = datastore.prepare(query);
+    return results.asList(FetchOptions.Builder.withLimit(10)).get(0);
+  }
+
+  /** This should only be called each time a new user makes an accout. */
+  public void createUserStats() { 
+     List<Key> items = new ArrayList<>();
+
+     Entity userStats = new Entity("UserStats");
+     userStats.setProperty("UserId", userId);
+     userStats.setProperty("Items", items);
+     datastore.put(userStats);
+  }
+
+  /** 
+   * Updates the Items list in the UserStats Entity in datastore
+   * corresponding to this user.
+   * @param items - Non-null list of item Keys to be added to the current Item list.
+   */
+  public void updateUserStats(List<Key> newItems) {
+    Entity userStats = retreiveUserStats();    
+    List<Key> items = (List<Key>) userStats.getProperty("Items");
+    if (items == null) {
+      items = newItems;
+    } else {
+      items.addAll(newItems);
+    }
+    userStats.setProperty("Items", items);
+    datastore.put(userStats);
+  }
+ 
+  /** 
+   * Copmiles the spending using the Item list found in this user's
+   * UserStats Entity in datastore.
+   * TODO: Allow for various time periods (only calculates weekly aggregate now).
+   * @return A map relating a time period to the spending in that time period.
+   */
+  public Map<String, String> aggregateUserData() { 
+    Entity userStats = retreiveUserStats();
+    List<Key> items = (List<Key>) userStats.getProperty("Items");
+
+    items.sort(SORT_BY_DATE);
+
+    return calculateWeeklyTotal(items);
+  }
+
+  /**
+   * Creates a String-Intger map relating weekly periods to spending.
+   * @param items - A list of {@code Key} objects that reference Item entities
+                  - in the datastore.
+   * @return Creates a map with keys for each ending day of a weekly period and  
+   *         values for the total spending during that period. 
+   */
+  public Map<String, String> calculateWeeklyTotal(List<Key> items)  {
+    Map<String, String> weeklyTotals = new HashMap<String, String>();
+    LocalDate currentEndOfWeek;
+    try {
+      currentEndOfWeek = getEndOfWeek(LocalDate.parse((String) datastore.get(items.get(0))
+                                          .getProperty("date"), dateFormatter));
+    } catch(EntityNotFoundException e) {
+      System.err.println("Error: Entity could not be located");
+      return null;
```

Comment:
Generally we want to avoid nulls in all situations unless absolutely necessary. They tend to be error-prone, difficult to debug, and require more boilerplate for bookeeping. Instead let's return an empty map and the caller can still operate without failing if they don't have a null check

Reply:
Done.